storage/local/storage.go

```diff
@@ -616,6 +609,54 @@ func (s *memorySeriesStorage) Append(sample *model.Sample) {
 	s.incNumChunksToPersist(completedChunksCount)
 }
 
+// Throttled implements Storage.
+func (s *memorySeriesStorage) Throttled() bool {
+	if s.getNumChunksToPersist() > s.maxChunksToPersist ||
+		float64(atomic.LoadInt64(&numMemChunks)) > float64(s.maxMemoryChunks)*toleranceFactorMemChunks {
+		s.throttled <- struct{}{}
+		return true
+	}
+	return false
+}
+
+// logThrottling handles logging of throttled events and has to be started as a
+// goroutine. It stops once s.loopStopping is closed.
+//
+// Logging strategy: Whenever Throttle() is called and returns true, an signal
+// is sent to s.throttled. If that happens for the first time, an Error is
+// logged that the storage is now throttled. As long as signals continues to be
+// sent via s.throttled at least once per minute, nothing else is logged. Once
+// no signal has arrived for a minute, an Info is logged that the storage is not
+// throttled anymore. This resets things to the initial state, i.e. once a
+// signal arrives again, the Error will be logged again.
+func (s *memorySeriesStorage) logThrottling() {
+	timer := time.NewTimer(time.Minute)
+	timer.Stop()
+
+	for {
+		select {
+		case <-s.throttled:
+			if !timer.Reset(time.Minute) {
+				log.
+					With("chunksToPersist", s.getNumChunksToPersist()).
+					With("maxChunksToPersist", s.maxChunksToPersist).
+					With("memoryChunks", atomic.LoadInt64(&numMemChunks)).
+					With("maxToleratedMemChunks", int(float64(s.maxMemoryChunks)*toleranceFactorMemChunks)).
+					Error("Storage is throttled. Scrapes and rule evaluations will be skipped.")
```

Comment:
As discussed yesterday, I'm not fully sure this is the place to log this "state" – but it's certainly the most convenient one implementation-wise right now.

If we have scrape intervals of greater than 1 minute this logging will be flapping. We don't really need to inform whether it stopped (since it's not an actual state anyway and the stopping message is unreliable as it depends on another call to `Throttle()`).
If the minutely logging goes away, that's enough to know in my opinion.

We also don't have to send on the channel on each call to `Throttle()` (might be a few hundred times/second) and consume each item.

To make this simpler, how about `s.throttle = make(chan struct{}, 1)`.

Then, in `Throttled()` so there is no potential blocking if the server is busy:
```go
select {
case s.throttled <- struct{}{}:
default:
    // Signal already pending.
}
```

In the consuming loop:
```go
timer := time.NewTimer(0)

for range timer.C {
    <-s.throttled
    timer.Reset(time.Minute)

    // Log the things.
}
```